lib/fs.js

```diff
@@ -204,22 +204,52 @@ function validateOffsetLengthWrite(offset, length, byteLength) {
   }
 }
 
-function validatePath(path, propName) {
-  let err;
+function validatePath(path, propName, assertFilename, callback) {
 
   if (propName === undefined) {
     propName = 'path';
   }
 
+  if (typeof assertFilename === 'function') {
+    callback = assertFilename;
+    assertFilename = false;
+  }
+
+  if (typeof propName === 'function') {
+    callback = propName;
+    propName = 'path';
+    assertFilename = false;
+  }
+
+  if (typeof propName === 'boolean') {
+    propName = 'path';
+    assertFilename = false;
+  }
+
+  const handle = (err) => {
+    if (typeof callback !== 'function') {
+      return handleError(err);
+    } else {
+      return handleError(err, callback);
+    }
+
+  };
+
   if (typeof path !== 'string' && !isUint8Array(path)) {
-    err = new errors.TypeError('ERR_INVALID_ARG_TYPE', propName,
-                               ['string', 'Buffer', 'URL']);
+    return handle(new errors.TypeError('ERR_INVALID_ARG_TYPE', propName,
+                                       ['string', 'Buffer', 'URL']));
   }
 
-  if (err !== undefined) {
-    Error.captureStackTrace(err, validatePath);
-    throw err;
+  if (assertFilename) {
+    if (typeof path === 'string' && path[path.length - 1] === '/') {
+      return handle(new errors.Error('ERR_INVALID_PATH_FORMAT', path));
+    }
+    if (isUint8Array(path) && path[path.length - 1] === 47) {
+      return handle(new errors.Error('ERR_INVALID_PATH_FORMAT', path));
```

Comment:
Note that this would be displayed as `Path 102,105,108,101,58,47,47,47,116,109,112,47,116,101,115,116,47 ended with an invalid suffix character`, we can either

1. Do not display the content in the path if it is a Uint8Array
2. Decode it as UTF8 and display it

Personally I prefer 1 even though we currently assume the buffer is encoded in UTF8 in the C++ layer. You can tweak the formatter in `internal/errors.js` and use a custom function instead of a string passed to `util.format` to do this.